Validate @on_load function in Elixir

fertapric opened the pull request:
This is an attempt to implement the check from #5800: `{undefined_on_load,{F,A}}`: function for `on_load` undefined.

These are my first steps with Erlang 🎉 , any feedback is more than appreciated :) 

During the development of this PR, I had some doubts regarding the visibility of the `@on_load` function. According to the [Erlang's documentation](http://erlang.org/doc/reference_manual/code_loading.html#on_load), "It is not necessary to export the function". (I'm currently learning the Elixir internals and what I'm about to say might not make sense, so please be bear with me 😅) I would think that public functions map to exported functions, and private functions to non-exported functions. If that's the case, this should work (as the function does not need to be public/exported):

```elixir
defmodule Example
  @on_load :foo

  defp foo do
    IO.puts("HELLO")
    :ok
  end
end
```

However, when executing the code, this error is returned:

```
warning: function foo/0 is unused
  lib/test.ex:4

** (CompileError) lib/test.ex:1: function foo/0 undefined
    (stdlib) lists.erl:1338: :lists.foreach/2
    (stdlib) erl_eval.erl:677: :erl_eval.do_apply/6
```

I wasn't sure if this was an expected behavior. I've assumed the current implementation was right, and this PR will make the compiler raise a `CompileError` is the function used in `@on_load` is private. I've also updated the documentation to reflect the visibility of the `on_load` function.

josevalim commented:
>  I would think that public functions map to exported functions, and private functions to non-exported functions. If that's the case, this should work (as the function does not need to be public/exported).

Correct!. The reason why it doesn't work is because when you  define `:on_load` as a private function, we think it is never invoked, and then we completely removed it from the module, making compilation fail. We could fix this by letting the compiler know that the `@on_load` function has been used. But given that this never worked anyway, I would rather keep things as is for now.

josevalim commented on lib/elixir/src/elixir_module.erl:
```diff
@@ -92,6 +92,10 @@ compile(Line, Module, Block, Vars, E) ->
 
+    OnLoadAttribute = lists:keyfind(on_load, 1, Attributes),
+    OnLoadAttribute /= false andalso
+      validate_on_load_attribute(OnLoadAttribute, AllDefinitions, Unreachable, File, Line),
```
I would remove `OnLoadAttribute /= false` and check for false inside `validate_on_load_attribute`.

josevalim commented on lib/elixir/src/elixir_module.erl:
```diff
@@ -155,6 +159,16 @@ validate_inlines([Inline | Inlines], Defs, Unreachable, Acc) ->
+    false ->
+      elixir_errors:form_error([{line, Line}], File, ?MODULE, {undefined_on_load, Def});
+    _ ->
+      lists:member(Def, Unreachable) andalso
```
Because you are looking up on unreachable, it means this code will work:

```elixir
defmodule Foo do
  @on_load :baz
  def bar, do: baz
  defp baz, do: :ok
end
```

So I think we should rather check the result returned by `lists:keyfind/3`. One of the elements in the tuple should say if it is `def`, `defmacro`, `defp` or `defmacrop`.

josevalim commented:
Great job, I have added some comments!

fertapric commented:
I've updated the PR with a few changes:

* As suggested [here](https://github.com/elixir-lang/elixir/pull/7973#discussion_r204254404), the check for `false` is now made in the `validate_on_load_attribute` function.
* As pointed out [here](https://github.com/elixir-lang/elixir/pull/7973#discussion_r204254474), it now uses the definition to check that the `@on_load` function was implemented as a public function.
* I've changed the error when the `@on_load` attribute is defined with the wrong kind. That should cover the cases of `defp`, `defmacro` and `defmacrop`.
* I've updated the documentation to reflect the difference between the requirements of the `@on_load` function and its behavior.

josevalim commented on lib/elixir/test/elixir/kernel/errors_test.exs:
```diff
@@ -719,6 +719,30 @@ defmodule Kernel.ErrorsTest do
+
+  test "wrong kind for @on_load attribute" do
+    # Capture warning: function foo/0 is unused
+    ExUnit.CaptureIO.capture_io(:stderr, fn ->
```
If we do this, the testcase cannot be async. So let's define a public function that calls `defp foo` so we don't need to capture the warning.

fertapric replied:
Good catch! I completely forgot about that!

fertapric replied:
Done in https://github.com/elixir-lang/elixir/pull/7973/commits/86c46d744f560c8f3f785c6c0788174778d5474b

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
